n3fit/src/n3fit/model_trainer.py

```diff
@@ -524,11 +524,25 @@ def _generate_observables(
         log.info("Generating layers")
 
         # Now we need to loop over all dictionaries (First exp_info, then pos_info and integ_info)
-        for exp_dict in self.exp_info:
+        for index, exp_dict in enumerate(self.exp_info[0]):
             if not self.mode_hyperopt:
                 log.info("Generating layers for experiment %s", exp_dict["name"])
 
-            exp_layer = model_gen.observable_generator(exp_dict)
+            # Stacked tr-vl mask array for all replicas for this dataset
+            replica_masks = np.stack([e[index]["trmask"] for e in self.exp_info])
+            training_data = np.stack([e[index]["expdata"].flatten() for e in self.exp_info])
+            validation_data = np.stack([e[index]["expdata_vl"].flatten() for e in self.exp_info])
+            invcovmat = np.stack([e[index]["invcovmat"] for e in self.exp_info])
+            invcovmat_vl = np.stack([e[index]["invcovmat_vl"] for e in self.exp_info])
```

Comment:
Please avoid these list comprehensions. Since they are all loops over self.exp_info which is also the outside loop. It's really not clear what is happening here.

Reply:
The exp_info object is a list over experiments, where each entry is a list over replicas, where each entry is a dictionary with strings as keys ("expdata, "expdata_vl", ...) and numpy arrays as values. For the multi-replica observables, we need to collect the replica arrays and stack them.

I will create a separate loop to do this.

Reply:
No, I guess the loop cannot be avoided. I meant that it would be much clearer to have an outside loop filling in these lists instead of a list comprehension so that it is clear that they are all the same information (and that it is a double loop on self.exp_info)

At this point the performance is not important, better to do a few extra steps so everything is clearer. The performance is relevant later on the fit.

Reply:
I fixed it in  de3b55c